util/progress/metricwriter.go

```diff
@@ -0,0 +1,150 @@
+package progress
+
+import (
+	"context"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/docker/buildx/util/metricutil"
+	"github.com/moby/buildkit/client"
+	"github.com/opencontainers/go-digest"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+type metricWriter struct {
+	recorders []metricRecorder
+	attrs     attribute.Set
+}
+
+func newMetrics(mp metric.MeterProvider, attrs attribute.Set) *metricWriter {
+	meter := metricutil.Meter(mp)
+	return &metricWriter{
+		recorders: []metricRecorder{
+			newLocalSourceTransferMetricRecorder(meter, attrs),
+		},
+		attrs: attrs,
+	}
+}
+
+func (mw *metricWriter) Write(ss *client.SolveStatus) {
+	for _, recorder := range mw.recorders {
+		recorder.Record(ss)
+	}
+}
+
+type metricRecorder interface {
+	Record(ss *client.SolveStatus)
+}
+type (
+	localSourceTransferState struct {
+		// Attributes holds the attributes specific to this context transfer.
+		Attributes attribute.Set
+
+		// LastTransferSize contains the last byte count for the transfer.
+		LastTransferSize int64
+	}
+	localSourceTransferMetricRecorder struct {
+		// BaseAttributes holds the set of base attributes for all metrics produced.
+		BaseAttributes attribute.Set
+
+		// State contains the state for individual digests that are being processed.
+		State map[digest.Digest]*localSourceTransferState
+
+		// TransferSize holds the metric for the number of bytes transferred.
+		TransferSize metric.Int64Counter
+
+		// Duration holds the metric for the total time taken to perform the transfer.
+		Duration metric.Float64Counter
+	}
+)
+
+func newLocalSourceTransferMetricRecorder(meter metric.Meter, attrs attribute.Set) *localSourceTransferMetricRecorder {
+	mr := &localSourceTransferMetricRecorder{
+		BaseAttributes: attrs,
+		State:          make(map[digest.Digest]*localSourceTransferState),
+	}
+	mr.TransferSize, _ = meter.Int64Counter("source.local.transfer.io",
+		metric.WithDescription("Measures the number of bytes transferred between the client and server for the context."),
+		metric.WithUnit("By"))
+
+	mr.Duration, _ = meter.Float64Counter("source.local.transfer.time",
+		metric.WithDescription("Measures the length of time spent transferring the context."),
+		metric.WithUnit("ms"))
```

Comment:
These names are based on this section of the instrument naming document for OTEL: https://opentelemetry.io/docs/specs/semconv/general/metrics/#instrument-naming